openmc/deplete/integrator/abc.py

```diff
@@ -0,0 +1,268 @@
+            res = operator.prev_res[-1]
+            if res.data.shape[0] != self._N_STAGES:
+                raise ValueError(
+                    "{} incompatible with previous restart calculation. "
+                    "Previous scheme used {} intermediate solutions, while "
+                    "this uses {}".format(
+                        self.__class__.__name__, res.data.shape[0],
+                        self._N_STAGES))
+        self.operator = operator
+        self.chain = operator.chain
+        if not isinstance(timesteps, Iterable):
+            self.timesteps = [timesteps]
+        else:
+            self.timesteps = timesteps
+        if power is None:
+            if power_density is None:
+                raise ValueError("Either power or power density must be set")
+            if not isinstance(power_density, Iterable):
+                power = power_density * operator.heavy_metal
+            else:
+                power = [p * operator.heavy_metal for p in power_density]
+
+        if not isinstance(power, Iterable):
+            # Ensure that power is single value if that is the case
+            power = [power] * len(self.timesteps)
+        elif len(power) != len(self.timesteps):
+            raise ValueError(
+                "Number of time steps != number of powers. {} vs {}".format(
+                    len(self.timesteps), len(power)))
+
+        self.power = power
+
+    @abstractmethod
+    def __call__(self, conc, rates, dt, power, i):
+        """Perform the integration across one time step
+
+        Parameters
+        ----------
+        conc : numpy.ndarray
+            Initial concentrations for all nuclides in [atom]
+        rates : openmc.deplete.ReactionRates
+            Reaction rates from operator
+        dt : float
+            Time in [s] for the entire depletion interval
+        power : float
+            Power of the system [W]
+        i : int
+            Current depletion step index
+
+        Returns
+        -------
+        proc_time : float
+            Time spent in CRAM routines for all materials
+        conc_list : list of numpy.ndarray
+            Concentrations at each of the intermediate points with
+            the final concentration as the last element
+        op_results : list of openmc.deplete.OperatorResult
+            Eigenvalue and reaction rates from intermediate transport
+            simulations
+        """
+
+    @property
+    @abstractmethod
+    def _N_STAGES(self):
+        """Number of intermediate transport solutions
+
+        Needed to ensure schemes are consistent with restarts
+        """
+
+    def __iter__(self):
+        return zip(self.timesteps, self.power)
+
+    def __len__(self):
+        return len(self.timesteps)
+
+    def _get_bos_data_from_openmc(self, step_index, step_power, bos_conc):
+        x = deepcopy(bos_conc)
+        res = self.operator(x, step_power)
+        self._write_statepoint(step_index)
```

Comment:
Agree. My hesitation with `Operator.__call__` is that some routines call that a lot during the integration. Maybe provide a method that is just `write_bos_data` on the base class, that the `Operator` uses to write statepoint data?

Reply:
Resolution: 1) method renamed to `_get_bos_data_from_operator`  and 2) added a `write_bos_data` abstract method to `TransportOperator`. This method accepts a single integer of the depletion index, and the `Operator` uses this to write the statepoint data.

This is nice because now our `DummyOperator` used in testing doesn't have to make calls to the C API, as it's `write_bos_data` method can be empty